src/caffe/parallel.cpp

```diff
@@ -391,15 +392,14 @@ void P2PSync<Dtype>::run(const vector<int>& gpus) {
   LOG(INFO)<< "GPUs pairs " << s.str();
 
   SolverParameter param(solver_->param());
-  vector<shared_ptr<P2PSync<Dtype> > > syncs(gpus.size());
 
   // Build the GPU tree by finding the parent for each solver
   for (int attempts = 0; attempts < pairs.size(); ++attempts) {
     for (int i = 1; i < pairs.size(); ++i) {
-      if (!syncs[i].get()) {
+      if (!syncs->at(i).get()) {
```

Comment:
This is not my point. While it's clear that we've switched to a pointer type, we've _also_ switched from calling `operator[]` to the `at` method. (You could just as well have written `(*syncs)[i]` in the new code, or `syncs.at(i)` in the old code.) Please note that these are not the same; the former does not perform bounds checking, while the latter does. 

Now presumably bounds checking _is_ the right thing to do here; there's not going to be a performance issue here, and the indexing is not very obviously within bounds.

Furthermore, we do indeed disallow non-`const` references. However, the right thing to do here might be to pass a `const` reference. Even though we use pointer types for output arguments, note that the output arguments here are really the `P2PSync` elements of the vector, not the vector itself (e.g., it's not okay for `Prepare` to add or remove elements from this vector, even though that's allowed in the current patch). See #945 for a lot more discussion on this point.

So I'd recommend changing the type of `syncs` to a `const` reference, but _keeping_ the calls to `at`, if you think that makes sense.